packages/fiber/src/core/stages.ts

```diff
@@ -0,0 +1,109 @@
+import { MutableRefObject } from 'react'
+import { StoreApi, UseBoundStore } from 'zustand'
+import { RootState } from './store'
+
+export interface UpdateCallback {
+  (state: RootState, delta: number, frame?: THREE.XRFrame | undefined): void
+}
+
+export type UpdateCallbackRef = MutableRefObject<UpdateCallback>
+type Store = UseBoundStore<RootState, StoreApi<RootState>>
+export type UpdateSubscription = { ref: UpdateCallbackRef; store: Store }
+
+export type FixedStageOptions = { fixedStep?: number; maxSubSteps?: number }
+
+export class Stage {
+  name: string
+  subscribers: UpdateSubscription[]
+
+  constructor(name: string) {
+    this.name = name
+    this.subscribers = []
+  }
+
+  frame(delta: number, frame?: THREE.XRFrame | undefined) {
+    const subs = this.subscribers
+
+    for (let i = 0; i < subs.length; i++) {
+      subs[i].ref.current(subs[i].store.getState(), delta, frame)
+    }
+  }
+
+  add(ref: UpdateCallbackRef, store: Store) {
+    this.subscribers = [...this.subscribers, { ref, store }]
+
+    return () => {
+      this.subscribers = this.subscribers.filter((sub) => {
+        return sub.ref !== ref
+      })
+    }
+  }
+}
+
+const FPS_50 = 1 / 50
+
+export class FixedStage extends Stage {
```

Comment:
If this is supposed to be exposed to public, as I assume, I'd add a jsdoc-style comment that explains what it does and when it's used